### PR TITLE
Int16 use

### PR DESCRIPTION
**Pull request by Saksham20**
changes to convert raw data into Electrical SEries storing channel conversion values separately. Raw datasets are now int16.

### CODE REVIEW

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py [hunk 1]
```diff
@@ -388,15 +399,25 @@ def create_acquisition(self):
...
                     if i['name'] in ['raw.lf', 'raw.ap']:
                         for j, probes in enumerate(range(self.no_probes)):
+                            if i['data'][j].shape[1] > self.one_data.data_attrs_dump['electrode_table_length'][j]:
+                                if 'channels.rawInd' in self.one_data.loaded_datasets:
+                                    channel_idx = self.one_data.loaded_datasets['channels.rawInd'][j].data
+                                else:
+                                    warnings.warn('could not find channels.rawInd')
+                                    break
+                            else:
+                                channel_idx = np.array(range(i['data'][j].shape[1]))
```
Comment (bendichter) [inline ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py:406]:
```suggestion
                                channel_idx = np.arange(i['data'][j].shape[1]))
```

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py [hunk 2]
```diff
@@ -388,15 +399,25 @@ def create_acquisition(self):
...
                             self.nwbfile.add_acquisition(
-                                TimeSeries(
+                                ElectricalSeries(
                                     name=i['name'] + '_' + self.nwb_metadata['Probes'][j]['name'],
-                                    starting_time=i['timestamps'][j][0, 1],
+                                    starting_time=np.abs(np.round(i['timestamps'][j][0, 1],2)),
```
Comment (bendichter) [inline ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py]: Why are you rounding this?

Comment (Saksham20) [reply ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py]: starting times for some datasets are of the order of -1e-5, so I round to nearest 2 decimals.

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py [hunk 3]
```diff
@@ -296,10 +290,22 @@ def create_timeseries_ecephys(self):
...
+                        probe_no = [j for j in range(self.no_probes)
+                                    if self.nwb_metadata['Probes'][j]['name'] in data_names[data_idx]][0]
+                        if data.shape[1] > self.one_data.data_attrs_dump['electrode_table_length'][probe_no]:
+                            if 'channels.rawInd' in self.one_data.loaded_datasets:
+                                channel_idx = self.one_data.loaded_datasets['channels.rawInd'][probe_no].data
+                            else:
+                                warnings.warn('could not find channels.rawInd')
+                                break
+                        else:
+                            channel_idx = np.array(range(data.shape[1]))
```
Comment (bendichter) [inline ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py:302]:
```suggestion
                            channel_idx = np.arange(data.shape[1])
```

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py [hunk 4]
```diff
@@ -296,10 +290,22 @@ def create_timeseries_ecephys(self):
...
+                        mod.add(ElectricalSeries(name=data_names[data_idx],
+                                                 description=i['description'],
+                                                 timestamps=i['timestamps'][
+                                                     timestamps_names.index(data_names[data_idx])],
+                                                 data=data[:, channel_idx],
```
Comment (bendichter) [inline ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py:307]: Wouldn't `channel_idx` always span the 2nd dim of `data`? Why not just have `data=data`?

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py [hunk 5]
```diff
@@ -388,15 +396,27 @@ def create_acquisition(self):
...
+                                    warnings.warn('could not find channels.rawInd')
+                                    break
+                            else:
+                                channel_idx = np.arange(i['data'][j].shape[1]))
                             self.nwbfile.add_acquisition(
-                                TimeSeries(
+                                ElectricalSeries(
                                     name=i['name'] + '_' + self.nwb_metadata['Probes'][j]['name'],
-                                    starting_time=i['timestamps'][j][0, 1],
+                                    starting_time=np.abs(np.round(i['timestamps'][j][0, 1], 2)),
```
Comment (bendichter) [inline ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py:410]: why are you rounding?

Comment (Saksham20) [reply ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py:410]: https://github.com/catalystneuro/IBL-to-nwb/pull/15#discussion_r520618282

File: ibl_to_nwb/AlyxToNWB/io_tools.py [hunk 6]
```diff
@@ -1,24 +1,27 @@
...
     ----------
-    datasetview_obj: DatasetView
+    datasetview_obj: np.array
         2-D array to iteratively write to nwb.
+    channel_ids: np.array
+        channel numbers to store
     """
-
+    if channel_ids is None:
+        channel_ids = np.array(range(datasetview_obj.shape[1]))
```
Comment (bendichter) [inline ibl_to_nwb/AlyxToNWB/io_tools.py:22]:
```suggestion
```

File: ibl_to_nwb/AlyxToNWB/io_tools.py [hunk 7]
```diff
@@ -1,24 +1,27 @@
...
     """
     Generator to return a row of the array each time it is called.
     This will be wrapped with a DataChunkIterator class.
 
     Parameters
     ----------
-    datasetview_obj: DatasetView
+    datasetview_obj: np.array
         2-D array to iteratively write to nwb.
+    channel_ids: np.array
```
Comment (bendichter) [inline ibl_to_nwb/AlyxToNWB/io_tools.py:18]:
```suggestion
    channel_ids: np.array of ints, optional
```

File: ibl_to_nwb/AlyxToNWB/io_tools.py [hunk 8]
```diff
@@ -1,24 +1,27 @@
...
-def iter_datasetview(datasetview_obj):
+def iter_datasetview(datasetview_obj, channel_ids=None):
     """
     Generator to return a row of the array each time it is called.
     This will be wrapped with a DataChunkIterator class.
 
     Parameters
     ----------
-    datasetview_obj: DatasetView
+    datasetview_obj: np.array
```
Comment (bendichter) [inline ibl_to_nwb/AlyxToNWB/io_tools.py:16]:
```suggestion
    datasetview_obj: array-like
```

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py (hunk 4 above)
Comment (Saksham20) [reply ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py:307]: these channels_idx are a subset of the total channels in the raw dataset. : `len(channel_idx)<=data.shape[1]`

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py (hunk 2 above)
Comment (bendichter) [reply ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py]: Then comment that

File: ibl_to_nwb/AlyxToNWB/io_tools.py [hunk 9]
```diff
@@ -1,24 +1,27 @@
...
         2-D array to iteratively write to nwb.
+    channel_ids: np.array
+        channel numbers to store
     """
-
+    if channel_ids is None:
+        channel_ids = np.array(range(datasetview_obj.shape[1]))
     for i in range(datasetview_obj.shape[0]//700):
-        curr_data = np.squeeze(datasetview_obj[i:i + 1])
+        curr_data = np.squeeze(datasetview_obj._raw[i][channel_ids])
```
Comment (bendichter) [inline ibl_to_nwb/AlyxToNWB/io_tools.py:22]:
```suggestion
        curr_data = np.squeeze(datasetview_obj._raw[i, channel_ids])
```

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py (hunk 4 above)
Comment (bendichter) [reply ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py:307]: oh, I see

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py [hunk 10]
```diff
@@ -296,10 +290,22 @@ def create_timeseries_ecephys(self):
...
+                        probe_no = [j for j in range(self.no_probes)
+                                    if self.nwb_metadata['Probes'][j]['name'] in data_names[data_idx]][0]
+                        if data.shape[1] > self.one_data.data_attrs_dump['electrode_table_length'][probe_no]:
+                            if 'channels.rawInd' in self.one_data.loaded_datasets:
+                                channel_idx = self.one_data.loaded_datasets['channels.rawInd'][probe_no].data
+                            else:
+                                warnings.warn('could not find channels.rawInd')
+                                break
+                        else:
+                            channel_idx = np.arange(data.shape[1]).astype('int')
```
Comment (bendichter) [inline ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py:302]:
```suggestion
                            channel_idx = None
```

File: ibl_to_nwb/AlyxToNWB/io_tools.py (hunk 9 above)
Comment (Saksham20) [reply ibl_to_nwb/AlyxToNWB/io_tools.py:22]: this seems to give an error with their dataset. It seems to be stored an array of arrays on their end

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py (hunk 5 above)
Comment (bendichter) [reply ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py:410]: Please comment that

File: ibl_to_nwb/AlyxToNWB/io_tools.py [hunk 11]
```diff
@@ -1,24 +1,25 @@
...
     Parameters
     ----------
-    datasetview_obj: DatasetView
+    datasetview_obj: array-like
         2-D array to iteratively write to nwb.
+    channel_ids: np.array of ints, optional
+        channel numbers to store
     """
-
     for i in range(datasetview_obj.shape[0]//700):
```
Comment (bendichter) [inline ibl_to_nwb/AlyxToNWB/io_tools.py:21]: What is this 7000 for?

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py [hunk 12]
```diff
@@ -264,17 +259,16 @@ def create_electrode_table_ecephys(self):
...
-                                                                              region=list(range(sum(
-                                                                                  self.one_data.data_attrs_dump[
-                                                                                      'electrode_table_length']))),
-                                                                              description='AllProbes')
+        self.probe_dt_region = [
+            self.nwbfile.create_electrode_table_region(region=list(range(self.one_data.data_attrs_dump[
+                                                                             'electrode_table_length'][
+                                                                             j])),
+                                                       description=i['name'])
+            for j, i in enumerate(self.nwb_metadata['Probes'])]
```
Comment (bendichter) [inline ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py:267]: Can you fix the indents on this? I can't read it

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py [hunk 13]
```diff
@@ -264,17 +259,16 @@ def create_electrode_table_ecephys(self):
...
+        self.probe_dt_region_all = self.nwbfile.create_electrode_table_region(region=list(range(sum(
+            self.one_data.data_attrs_dump[
+                'electrode_table_length']))),
+            description='AllProbes')
```
Comment (bendichter) [inline ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py:271]:
```suggestion
        self.probe_dt_region_all = self.nwbfile.create_electrode_table_region(
            region=list(range(sum(self.one_data.data_attrs_dump['electrode_table_length']))),
            description='AllProbes')
```

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py (hunk 2 above)
Comment (Saksham20) [reply ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py]: You mean remove that arg? I dont want to do that as some datasets might have a larger positive starting-time

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py [hunk 14]
```diff
@@ -388,15 +396,27 @@ def create_acquisition(self):
                 else:
                     if i['name'] in ['raw.lf', 'raw.ap']:
                         for j, probes in enumerate(range(self.no_probes)):
```
Comment (bendichter) [inline ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py:398]: `i` and `j`, etc. are ok for small for loops where the meaning if obvious, but they are confusing for large ones like this. Can you please rename these countervariables to things like `dataset_num` `probe_num` or something like that?

File: ibl_to_nwb/AlyxToNWB/io_tools.py (hunk 9 above)
Comment (bendichter) [reply ibl_to_nwb/AlyxToNWB/io_tools.py:22]: I'm confused about the form of the dataset. I've never seen this `_raw` thing before. Is this really an np.array as the docstring suggests?

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py (hunk 2 above)
Comment (bendichter) [reply ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py]: No, I mean add "starting times for some datasets are of the order of -1e-5, so I round to nearest 2 decimals" as a comment. You have very few comments and it is difficult for me to understand what you are doing in some places

**Comment by bendichter** [comment]
@Saksham20 please fix merge conflicts

File: ibl_to_nwb/AlyxToNWB/io_tools.py (hunk 11 above)
Comment (Saksham20) [reply ibl_to_nwb/AlyxToNWB/io_tools.py]: had added that to test a subset of raw_Datasets, forgot to remove that while commit. Removed now

File: ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py (hunk 14 above)
Comment (Saksham20) [reply ibl_to_nwb/AlyxToNWB/alyx_to_nwb_converter.py:396]: i'll merge this and create a separate pr for renaming and adding comments where it would help

**Pull request merged**